staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -31,6 +31,77 @@ type DelayingInterface interface {
+	AddAfterWithOptions(item interface{}, duration time.Duration, opts AddAfterOptions)
+	// ForgetDelayed will remove an item from the delayed/waiting queue - it won't cancel active queued items
+	// for them you are still expected to call Done()
+	ForgetDelayed(item interface{}, opts ForgetDelayedOptions)
+}
+
+// AddOptions allows callers to choose from different behaviours when calling AddWithOptions().
+type AddOptions struct {
+	// Waiting specifies how you want the immediate Add to interplay with the existing queue state and
+	// waiting items.
+	Waiting AddVsWaitingBehaviour
+	// Synchronous, when true, instructs AddWithOptions to wait until the item has been proceesed
+	// into the 'active' queue (or dropped) before returning.  Useful for clients/tests that need to
+	// inspect the state of the queue following the add.
+	Synchronous bool
+}
+
+// AddVsWaitingBehaviour defines different choices of how immediate adds interact with waiting items.
+type AddVsWaitingBehaviour int
+
+const (
+	// PreemptWaiting, the default, drops a waiting item when the same item is queued to the 'active' queue.
+	PreemptWaiting AddVsWaitingBehaviour = iota
+	// DropIfWaiting, drops and incoming add when the same item is already in the waiting queue.
+	// This helps us respect rate-limiters and back-offs because we can choose to prioritise the waiting item
+	// over new adds or retries.
+	DropIfWaiting
+	// IgnoreWaiting, will ignore items in the waiting queue and immediately add the item without checking the waiting
+	// queue. Adds made with this option will be slightly faster because they by-pass the waiting check but will result
+	// in superfluous re-executions of the same item (see https://github.com/kubernetes/kubernetes/issues/112325)
+	IgnoreWaiting
+)
+
+// AddAfterOptions toggle implementation specific options for use with calls to AddAfterWithOptions().
+type AddAfterOptions struct {
+	// Existing specifies how you want the waitingLoop to insert items that are already waiting.
+	Existing AlreadyWaitingBehaviour
+	// IgnoreActive, when true, will add a delayed item to the 'waiting' queue when the same item is in the 'active' queue.
+	// The default behaviour is to drop a request to queue a delayed item when the same item is actively queued, because
+	// the actively queued item will soon execute and could fail or succeed, either outcome could render the delayed execution
+	// unnecessary (not recommended - included for completeness and obscure use cases)
+	IgnoreActive bool
+	// Synchronous, when true, instructs AddAfterWithOptions to wait until the item has been proceesed
+	// into either the 'active' or 'waiting' queue (or dropped) before returning.  Useful for clients/tests that need to
+	// inspect the state of the queue following the add.
+	Synchronous bool
+}
+
+// AlreadyWaitingBehaviour defines different choices of how to handle a delayed add when there is already the same waiting 'item'.
+type AlreadyWaitingBehaviour int
```

Comment:
Can we merge this with the other Behavior enum? E.g. why is there both PreemptWaiting and TakeShorter?

Reply:
Thanks have now done so in the latest commit - one Add method - AddWithOptions and a single options type DelayingOptions.